src/rgw/rgw_bucket_layout.h

```diff
@@ -55,6 +55,7 @@ struct bucket_index_normal_layout {
   uint32_t num_shards = 1;
 
   BucketHashType hash_type = BucketHashType::Mod;
+  uint64_t reshard_gen = 0;
```

Comment:
@cbodley Thank you for your detailed review, I will work based on the comments as soon as possible. For this one, IMO, target_index->gen will clear when revert_target_layout() be called. So there will be duplicated index entries remaining after reshard failed, that can lead to redundant copys in a new reshard process. And more, if the duplicated entries is delete, and the same entry was written again in the between of the reshards, the dst index may be deleted wrongly. I do not try to change the revert_target_layout(), so add a new reshard_gen(added when reshard failed). But as you say, if trim_reshard_log_entries() is called by reshard failed automatically, I think target_index->gen may be enough.

Reply:
> Or for safety, retaining 'gen' in 'reshard logkey', making current_index->gen = target_index->gen firstly before clearing target_index in revert_target_layout(), so target_index->gen can be used.

we can't change `current_index.gen`, because we need that to locate the existing bucket index shard objects (their object names end in `.gen.shard-id`). if necessary, we could add a new member like `BucketLayout::next_reshard_gen` to help prevent us from reusing generation numbers in `target_index->gen`

> I think we do not even need 'gen' in 'reshard log key' anymore.

i like the sound of this, if we can guarantee that any existing log entries are cleared when we start a new reshard

Reply:
> i like the sound of this, if we can guarantee that any existing log entries are cleared when we start a new reshard
Another goal of next_reshard_gen occurs to me, as we discussed before: 
> if the block time is still too long, one more logrecord phase can be added. When dealing with initial record logs,
> do not blocking writes, instead, insert logs. And block writes in dealing with secord logs, this may just blocking
> writes for several seconds.
after the first logrecord phase, we do not change the logrecord state, but increase next_reshard_gen, meanwhile copying the duplicated index entries to dst shards by calling reshard_in_logrecord_state(old_next_reshard_gen). And in second round(block writes), only duplicated entries with new next_reshard_gen need to be copyed.

Reply:
multiple rounds of the logrecord stage is an interesting idea. but our existing use of `gen` is tied to the naming of bucket index shards, so should only change once per reshard

for the purposes of cls_rgw's reshard log, we could pass something other than a generation number to `cls_rgw_set_bucket_resharding()` for its namespacing of those log entries. that could be a `uint32_t round` that starts at 0 and increments each time we loop. but `round` could just be a local variable in the reshard logic, right? i don't think it needs to be persisted with the bucket info. we might also avoid needing separate namespaces if we can trim the log between rounds

i'd like to believe that a single round is sufficient. we can follow up with shard parallelism (discussed in https://github.com/ceph/ceph/pull/56757) to further reduce the duration of each stage

Reply:
> but round could just be a local variable in the reshard logic, right? 

Yep!

> we can follow up with shard parallelism (discussed in https://github.com/ceph/ceph/pull/56757) to further reduce the duration of each stage

In our experience, some clusters already have high ioutils for disks used by osds.  copying index from src to dst serially and duplicated written index entries in first stage will make more use of disks, parallelism in this stage may leads to slow ops of osds. So i think parallelism in second stage is more efficient and helpful.

> i'd like to believe that a single round is sufficient.

So, I will remove the `reshard_gen` from  bucket_index_normal_layout, and remove 'gen' from 'reshard log key'.